core/src/main/java/io/grpc/Contexts.java

```diff
@@ -149,4 +159,163 @@ public static Status statusFromCancelled(Context context) {
     }
     return status.withCause(cancellationCause);
   }
+
+  /**
+   * A utility for performing an operation inside of a
+   * {@link io.grpc.Context.CancellableContext CancellableContext}, and then cancelling the context
+   * after it is done.
+   *
+   * <p>Note: Because gRPC must be compatible with Java 1.6, this class does not support the
```

Comment:
We should be able to support the idiom even without breaking Java 6 compatibility. I thought extending Closeable sufficed because AutoCloseable is its super-interface.

I think we may want the ListenableFuture helper to also create the CancellableContext so the API works cleanly with http://errorprone.info/bugpattern/MustBeClosedChecker

Reply:
Added a new enclosing class (WithCancellableContext) that extends Closeable, and added `@MustBeClosed` to the factory method. From the javadocs, it looks like that annotation can only be used for things that are meant to be used for things initialized in a try-with-resources block. The CancellableContextApplier won't fit with that idiom.